### ISSUE TITLE
Compiler speed improvements with many selection subsets

### ISSUE THREAD

**Issue by csaluski** [tdhock/animint2#57]
While working on some visualizations I've run into some performance issues that seem to be related to subsetting large data sets with `showSelected` on several criteria. I've created a simplified example of the plot that I'm creating using synthetic data [here](https://gist.github.com/csaluski/91d47e262fac3e50572e6c71c5d7cd88). This code can be easily expanded to see what other variables may have an impact on compilation speed.

Profiling with profvis shows that the second system with the addition of groups takes almost twice as long to compile, although in my original code where I had performance issues it took even longer. A large majority of the time is spent in [`getCommonChunk`](https://github.com/tdhock/animint2/blob/6431aba10bc76adf5d0425a66eada21cba46f90f/R/z_animintHelpers.R#L742-L750), which I believe is called when the Geom objects are calculating how to optimize their data into chunks for file output.

It may be worth examining the performance characteristics as further subset options are added to figure out what the approximate runtime complexity is, so that future users can be aware of what performance considerations their plots have.

**Comment by csaluski** [comment]
See [this PR](https://github.com/tdhock/animint2/pull/8), some discussions about this previously, although it didn't focus on subsets increasing compilation time.

**Comment by Faye-yufan** [comment]
Hi @csaluski , I'm a little confused.
You mean compare ```grouped.scatter.plot``` with ```scatter. plot```, the additional subsets
```
geom_text(
    data = var.name.dt,
    aes(label = var.main.name),
    ...
  ) +
  geom_text(
    data = var.name.dt,
    aes(label = var.second.name),
    ...
  )
```
is what causes almost twice compilation time? 
And how do users interpret the approximate runtime complexity?

**Comment by csaluski** [comment]
I think I added the `geom_text`s to just slightly improve the visualization, they slightly increase the runtime but not by a large amount. What really causes the increase is the `clickSelects` in the grouped plot, see 
```
scatter.plot <- ggplot() +
  geom_point(
    data = combinations.data.table,
    aes(x = var.main.val, y = var.second.val),
    showSelected = c("var.main.name", "var.second.name")
  )

grouped.scatter.plot <- ggplot() +
  geom_point(
    data = combinations.data.table,
    aes(x = var.main.val, y = var.second.val, key = var.main.val),
    clickSelects = "i.group",
    showSelected = c("var.main.name", "var.second.name")
  ) 
```
I updated the Gist with one that shows only this difference, along with another that shows the original text labels. The `clickSelects` attribute forces the compiler to compute all of the different subsets of the `combinations.data.table` by `i.group`. 
I encourage you to run this in RStudio yourself to verify the results. If you don't have a full local install of RStudio set up you can run a Docker container for it with the command `docker run -d -p 8787:8787 -e DISABLE_AUTH=true -v ~/:/home/rstudio rocker/rstudio` and then connect to it on `localhost:8787`, this is how I run it. 

As for the runtime complexity, this would have to be something we would have to experiment with ourselves, and attempt to approximate the runtime from these experiments. This would let future users know ahead of time if their complicated plot systems would take a long time to compile. I don't think this aspect should be a high priority, but it would be interesting to see the results.

**Comment by Faye-yufan** [comment]
Inside `getCommonChunk`, it seems the `<Anonymous>` takes a big part of the time, which I think is stands for [this line](https://github.com/tdhock/animint2/blob/96ec0caf446c555ae847febdf3681de6a6525ee8/R/z_animintHelpers.R#L763) 
```
  all.nas <- sapply(built, function(x){all(is.na(x))})
```

So I tested different apply functions, `lapply()` somehow faster than `sapply()`, and the difference is negligible but I think is worth trying.

![image](https://user-images.githubusercontent.com/62975717/172467365-4e2ffdac-1db7-467a-8cdf-b8d90c2c65a4.png)

**Comment by csaluski** [comment]
Interesting results, it could definitely be worth looking at which version of apply is the best in each situation. However I don't think it's that specific instance of the function, as that's only one call over the data structure and not a loop, which should be a very small part of the runtime of the function, and it's not in quite the right order. In the flame graph I see that before the anonymous function chunk there's a call to `table`, then it's followed by a call to `split.data.frame`, which I believe places it in [this chunk](https://github.com/tdhock/animint2/blob/96ec0caf446c555ae847febdf3681de6a6525ee8/R/z_animintHelpers.R#L779-L792) of code. I'm not sure if there's a way to get more specific timing information from anonymous functions, so you may want to do some more research on the options `profvis` provides and maybe @tdhock can help as well.

**Comment by tdhock** [comment]
that anonymous function,  `all.nas <- sapply(built, function(x){all(is.na(x))})`
does not seem like it should take a lot of time, so I think you should maybe try to re-run or re-analyze the profile. It should tell you which line of code takes the most time.

**Comment by Faye-yufan** [comment]
Thanks to @csaluski created a simplified framework for ```getCommonChunk``` func and I made [the code](https://gist.github.com/csaluski/91d47e262fac3e50572e6c71c5d7cd88) run. 
But ```getCommonChunk``` seems to cost trivial time(100-150ms) in this version. Still looking into it.

**Comment by Faye-yufan** [comment]
I started rewriting ```getCommonChunk``` func with ```data.table``` and tested it with @csaluski 's example above in the past few days. 

It seems to get some improvements until now. I only modified some parts of the function and the time cost has already been reduced by around 15%. See the profile [here](https://bl.ocks.org/Faye-yufan/3f2129fbb9ea9961d9ce74b90158d60e), where ```animint2dir_dt()``` is the new func.

@tdhock Please let me know if it is workable.

**Comment by tdhock** [comment]
hi thanks for the info! 15% does not seem like much of an improvement. Are you sure you have identified the slowest parts of the code correctly?

**Comment by Faye-yufan** [comment]
@tdhock update on the code profile and improvement:
I used the RStudio profiling tool to look into which part of code slows `getCommonChunk()`, and I think it is [this line](https://github.com/tdhock/animint2/blob/1e7c2683edf74191a4976a46d7b334d9c3840cd5/R/z_animintHelpers.R#L785) as below:  
```R
built <- built[do.call(order, order.args),]
```
This is for sorting the `showSelected` columns which will be chunked on, so as to assign `group` based on a certain level of order.

When there are multiple showSelected variables assigned, worse case each of them contains many different values, the data.frame `built` will become very large because of the value combinations. As a result, the code above takes more time compared to building chunks with only one showSelected variable.

I tested by converting `built` to a data.table and used `setorderv()` instead. 
```
built <- data.table::setorderv(built, chunk.vars)
```
Still I used [this example](https://gist.github.com/csaluski/91d47e262fac3e50572e6c71c5d7cd88), and here is the latest [profiling result](https://bl.ocks.org/Faye-yufan/raw/ccef1dc9e778809262c30e1829c765af/), where the time cost of `animint2dir_dt`(with data.table dependency) is reduced almost by half when it comes to many selection subsets.

**Comment by tdhock** [comment]
2x speedup sounds good.

**Comment by tdhock** [comment]
what is the new slowest part after using the new _dt implementation?

**Comment by Faye-yufan** [comment]
As for `animint2dir`, it is the [`saveChunks()`](https://github.com/tdhock/animint2/blob/1e7c2683edf74191a4976a46d7b334d9c3840cd5/R/z_animintHelpers.R#L935-L949) inside `storeLayer`.  As for `getCommonChunk`, it is the [`split.x()`](https://github.com/tdhock/animint2/blob/1e7c2683edf74191a4976a46d7b334d9c3840cd5/R/z_animintHelpers.R#L901-L927) to split data.frame into recursive list of data.frame. Both of these are recursive functions.

**Comment by tdhock** [comment]
saveChunks may get speedups by using data.table::fwrite.
split.x may get speedups if you use split on a data table (currently it is frame right?)

**Comment by Faye-yufan** [comment]
Right, currently it is dataframe.
Will do, thank you for the info!

**Comment by Faye-yufan** [comment]
@tdhock The `data.table::fwrite` works pretty well on this [profile](https://bl.ocks.org/Faye-yufan/raw/1e89abe6b6035b2ea808b87daca36423/), with the total time cost of `storeLayer` reduce from 5670ms to 200ms in this case.
`split.x` does not seem to improve much via `data.table` in this size(<1000 rows) of dataset, but will speed up in a bigger dataset(>10000 rows)
```r
> DT = data.table(x1 = rep(letters[1:2], 600), x2 = rep(letters[3:5], 400), x3 = rep(letters[5:8], 300), y = rnorm(12)) 
> DT = DT[sample(.N)]
> DF = as.data.frame(DT)
> DT.L = data.table(x1 = rep(letters[1:2], 60000), x2 = rep(letters[3:5], 40000), x3 = rep(letters[5:8], 30000), y = rnorm(12)) 
> DT.L = DT.L[sample(.N)]
> DF.L = as.data.frame(DT.L)

## Time cost of small dataframe
> system.time({ split(DF[names(DF) != "x1"], DF["x1"], drop = TRUE) })
   user  system elapsed 
  0.001   0.000   0.001 
## Time cost of small data.table
> system.time({ split(DT, by = "x1", keep.by = FALSE, drop = TRUE) })
   user  system elapsed 
  0.003   0.001   0.007 
## Time cost of large dataframe
> system.time({ split(DF.L[names(DF.L) != "x1"], DF.L["x1"], drop = TRUE) })
   user  system elapsed 
  0.037   0.003   0.040 
## Time cost of large data.table
> system.time({ split(DT.L, by = "x1", keep.by = FALSE, drop = TRUE) })
   user  system elapsed 
  0.011   0.000   0.011 
```

**Comment by tdhock** [comment]
great thanks for the info. if you make those changes then what is the next thing that pops up as slowest?
for comparing timings of different codes as a function of data size N, you may try using my new atime package, which can make figures like this,
![table-vs-frame](https://user-images.githubusercontent.com/932850/180315852-f58f59d7-059d-4f4b-9c8f-d070966f3b52.PNG)
https://github.com/tdhock/atime/blob/main/vignettes/split.Rmd

**Comment by Faye-yufan** [comment]
Awesome, thanks for letting me know!
The next slowest in getCommonChunk is `varied.chunk()`, same situation as split.x, which I have updated with data.table in the recent commit. And this [for loop](https://github.com/tdhock/animint2/blob/0a2988ff02f73aab7ec6c3fb70d14439092bee27/R/z_animintHelpers.R#L842-L856) with `apply` is also slow, I should probably change it as well.
And parsePlot which calls `ggplot_build()` is also kind of slow, and I saw there is a TODO said to implement a compiler that without ggplot_build and does all of the relevant computations in animint code. Any comment on this?
